File: apps/mobile/lib/features/globe/services/globe_bridge.dart
import 'dart:convert';

import 'package:flutter_inappwebview/flutter_inappwebview.dart';

import 'package:mobile/features/globe/models/globe_models.dart';

typedef OnPinTapped = void Function(String contactId, double lat, double lng);
typedef OnClusterTapped = void Function(
    List<String> contactIds, double lat, double lng, int count);
typedef OnLocationSelected = void Function(double lat, double lng);
typedef OnGlobeReady = void Function(String version);

class GlobeBridge {
  InAppWebViewController? _controller;

  OnGlobeReady? onReady;
  OnPinTapped? onPinTapped;
  OnClusterTapped? onClusterTapped;
  OnLocationSelected? onLocationSelected;

  void attach(InAppWebViewController controller) {
    _controller = controller;
  }

  void detach() {
    _controller = null;
  }

  /// Handle messages coming from the globe JS.
  void handleMessage(String rawMessage) {
    final msg = jsonDecode(rawMessage) as Map<String, dynamic>;
    final type = msg['type'] as String?;
    final payload = msg['payload'] as Map<String, dynamic>? ?? {};

    switch (type) {
      case 'READY':
        onReady?.call(payload['version'] as String? ?? '');
      case 'PIN_TAPPED':
        onPinTapped?.call(
          payload['contactId'] as String,
          (payload['lat'] as num).toDouble(),
          (payload['lng'] as num).toDouble(),
        );
      case 'CLUSTER_TAPPED':
        onClusterTapped?.call(
          (payload['contactIds'] as List).cast<String>(),
          (payload['lat'] as num).toDouble(),
          (payload['lng'] as num).toDouble(),
          payload['count'] as int,
        );
      case 'LOCATION_SELECTED':
        onLocationSelected?.call(
          (payload['lat'] as num).toDouble(),
          (payload['lng'] as num).toDouble(),
        );
    }
  }

  /// Send a message to the globe JS.
  Future<void> _send(Map<String, dynamic> message) async {
    final json = jsonEncode(message);
    await _controller?.evaluateJavascript(
      source: 'window.GlobeBridge.receive($json);',
    );
  }

  Future<void> setPins(List<GlobePin> pins) => _send({
        'type': 'SET_PINS',
        'payload': {'pins': pins.map((p) => p.toJson()).toList()},
      });

  Future<void> setArcs(List<GlobeArc> arcs) => _send({
        'type': 'SET_ARCS',
        'payload': {'arcs': arcs.map((a) => a.toJson()).toList()},
      });

  Future<void> setTheme(String theme) => _send({
        'type': 'SET_THEME',
        'payload': {'theme': theme},
      });

  Future<void> flyTo({
    required double lat,
    required double lng,
    double? altitude,
    int? durationMs,
  }) =>
      _send({
        'type': 'FLY_TO',
        'payload': {
          'lat': lat,
          'lng': lng,
          if (altitude != null) 'altitude': altitude,
          if (durationMs != null) 'durationMs': durationMs,
        },
      });

  Future<void> highlightContact(String contactId) => _send({
        'type': 'HIGHLIGHT_CONTACT',
        'payload': {'contactId': contactId},
      });

  Future<void> setMode(String mode) => _send({
        'type': 'SET_MODE',
        'payload': {'mode': mode},
      });
}

File: packages/types/src/globe-bridge.ts
// ---------------------------------------------------------------------------
// Globe JS Bridge — Message Schema
// Shared between Web (iframe postMessage) and Flutter (WebView JS channel).
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Common
// ---------------------------------------------------------------------------

export interface GlobeCoords {
  lat: number;
  lng: number;
}

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

export interface GlobePin {
  id: string;
  name: string;
  avatarUrl: string;
  lat: number;
  lng: number;
}

/** Arc communication type — each type has a distinct dash pattern (WCAG 1.4.1). */
export type ArcType = 'call' | 'email' | 'meeting' | 'message';

export interface GlobeArc {
  id: string;
  startLat: number;
  startLng: number;
  endLat: number;
  endLng: number;
  /** Communication type — controls dash pattern (solid, short dash, long dash, dot). */
  type: ArcType;
  /** Interaction frequency — controls arc stroke width. */
  frequency: number;
}

export type GlobeTheme = 'light' | 'dark';
export type GlobeMode = 'explore' | 'select_location' | 'view_connections';

// ---------------------------------------------------------------------------
// Inbound messages  (parent → globe)
// ---------------------------------------------------------------------------

export interface FlyToMessage {
  type: 'FLY_TO';
  payload: GlobeCoords & {
    altitude?: number;
    durationMs?: number;
  };
}

export interface HighlightContactMessage {
  type: 'HIGHLIGHT_CONTACT';
  payload: {
    contactId: string;
  };
}

export interface SetThemeMessage {
  type: 'SET_THEME';
  payload: {
    theme: GlobeTheme;
  };
}

export interface SetModeMessage {
  type: 'SET_MODE';
  payload: {
    mode: GlobeMode;
  };
}

export interface SetPinsMessage {
  type: 'SET_PINS';
  payload: {
    pins: GlobePin[];
  };
}

export interface SetArcsMessage {
  type: 'SET_ARCS';
  payload: {
    arcs: GlobeArc[];
  };
}

export type GlobeInboundMessage =
  | FlyToMessage
  | HighlightContactMessage
  | SetThemeMessage
  | SetModeMessage
  | SetPinsMessage
  | SetArcsMessage;

// ---------------------------------------------------------------------------
// Outbound messages  (globe → parent)
// ---------------------------------------------------------------------------

export interface ReadyMessage {
  type: 'READY';
  payload: {
    version: string;
  };
}

export interface PinTappedMessage {
  type: 'PIN_TAPPED';
  payload: GlobeCoords & {
    contactId: string;
  };
}

export interface ClusterTappedMessage {
  type: 'CLUSTER_TAPPED';
  payload: GlobeCoords & {
    contactIds: string[];
    count: number;
  };
}

export interface LocationSelectedMessage {
  type: 'LOCATION_SELECTED';
  payload: GlobeCoords;
}

export type GlobeOutboundMessage =
  | ReadyMessage
  | PinTappedMessage
  | ClusterTappedMessage
  | LocationSelectedMessage;

// ---------------------------------------------------------------------------
// Union of all messages
// ---------------------------------------------------------------------------

export type GlobeBridgeMessage = GlobeInboundMessage | GlobeOutboundMessage;

// ---------------------------------------------------------------------------
// Type guards
// ---------------------------------------------------------------------------

export function isInboundMessage(msg: GlobeBridgeMessage): msg is GlobeInboundMessage {
  return ['FLY_TO', 'HIGHLIGHT_CONTACT', 'SET_THEME', 'SET_MODE', 'SET_PINS', 'SET_ARCS'].includes(msg.type);
}

export function isOutboundMessage(msg: GlobeBridgeMessage): msg is GlobeOutboundMessage {
  return ['READY', 'PIN_TAPPED', 'CLUSTER_TAPPED', 'LOCATION_SELECTED'].includes(msg.type);
}
